runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -986,13 +977,57 @@ private Instant minTargetAndGcTime(Instant target) {
       return target;
     }
 
+    @Override
+    public Timer withOutputTimestamp(Instant outputTimestamp) {
+      this.outputTimestamp = outputTimestamp;
+      return this;
+    }
+
+    /** Verifies that the time domain of this timer is acceptable for absolute timers. */
+    private void verifyAbsoluteTimeDomain() {
+      if (!TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
+        throw new IllegalStateException(
+            "Cannot only set relative timers in processing time domain." + " Use #setRelative()");
+      }
+    }
+
+    /**
+     *
+     *
+     * <ul>
+     *   Ensures that:
+     *   <li>Users can't set {@code outputTimestamp} for processing time timers.
+     *   <li>Event time timers' {@code outputTimestamp} is set before window expiration.
+     * </ul>
+     */
+    private void setAndVerifyOutputTimestamp() {
+      // Output timestamp is currently not supported in processing time timers.
+      if (outputTimestamp != null && !TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
+        throw new IllegalStateException("Cannot set outputTimestamp in processing time domain.");
+      }
+      // Output timestamp is set to the delivery time if not initialized by an user.
+      if (outputTimestamp == null) {
+        outputTimestamp = target;
+      }
+
+      if (TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
+        Instant windowExpiry = window.maxTimestamp().plus(allowedLateness);
+        checkArgument(
+            !outputTimestamp.isAfter(windowExpiry),
```

Comment:
After a lot of thought, I think this logic is incorrect. I believe that setting a timer after the expiration of the window is invalid, even if the output timestamp is within the window (for one thing, the GC timer will have already fired). Changing this will also invalidate the test in ParDoTest (it will fail because the output timestamp is late).

@kennknowles really need your input here.